### PR TITLE
Visit admin page command

### PR DESCRIPTION
**Pull request by cadic**
### Description of the Change

Adds new command `visitAdminPage`

### Checklist:

- [x] I have read the [**CONTRIBUTING**](/CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my change.
- [x] All new and existing tests passed.

### Changelog Entry

Added - visitAdminPage command

### Credits

Props @cadic

### CODE REVIEW

**Review by dinhtungdu — COMMENTED** [review]
@cadic I don't think we should do more than one task for a command. Currently, for the login, we're no longer check if the user is logged in or not, we just go to the login page and log the user in.

`visitAdminPage`can be used more than one time in a test case, so this will break the test if we use this command when users are already logged in.

I think it should be the testers' responsibility to `login` user first, then `visitAdminPage`. By simplifying the command, it's also easier for testers to write the tests case using our utilities.

**Comment by cadic** [comment]
@dinhtungdu in this case, the `visitAdminPage` command become pointless, since it will not differ from the existing `visit`. I suppose this PR could be closed then.

**Comment by dinhtungdu** [comment]
@cadic Rethink about this PR, I think we can do conditional testing here. The reason I don't want to do more than one task for a command is to avoid conditional testing which leads to flaky tests. However, we get some good reasons to apply the condition testing for this command (and maybe also the `login` command):

- We're 100% sure about the state of DOM in this case: we know how it looks like if the users are logged in or not.
- It saves developers many keystrokes.

So I'm sorry for my former incomplete review! Reopening this PR.

File: src/commands/visit-admin-page.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+ * cy.visitAdminPage('http://localhost:8889/hello-world/', 'author', 'password')
+ * ```
+ */
+export const visitAdminPage = (
+  page = 'index.php',
+  username = 'admin',
+  password = 'password'
+): void => {
+  cy.login(username, password);
+  if (/^http/.exec(page)) {
```
Comment (dinhtungdu) [inline src/commands/visit-admin-page.ts:32]: Because this command visits an admin page, why do we need to support the full URL?

**Review by dinhtungdu — COMMENTED** [review]
How about try visiting the admin page first, then checking if the browser is redirected to the `wp-login.php` page or not? If yes, then we only login in that case.

File: src/commands/visit-admin-page.ts (hunk 1 above)
Comment (cadic) [reply src/commands/visit-admin-page.ts:32]: Originally I was thinking about using this command to allow visiting any page on the website as logged-in user (not only /wp-admin/*)

**Comment by cadic** [comment]
@dinhtungdu with conditional testing approach, I see the following scenario:

1. Attempt to visit admin page
2. If not logged in yet, the login form will appear
3. Use `login` command to log in
4. After login, the browser will redirect to the target admin page

This approach will require conditional update for `login` command because right now it force to open login page which will loose the `redirect_to` URL param.

Also, with this approach it will be hard to switch users: if we already logged in as other user (f.e. Author), the target page might be opened successfully on step 1 or we receive "You cannot access this page" blocker instead of login page. So `username` and `password` parameters of `visitAdminPage` command become pointless.

**Comment by cadic** [comment]
Actually, I agree with your initial review, saying "one task per command". It will help to isolate each command from being dependent on the state.

Assuming that tests are written with best practices, we expect that session and state are reset before each test, and it's testers responsibility. https://docs.cypress.io/guides/references/best-practices

**Comment by cadic** [comment]
With all above, the resulting `visitAdminPage` become just a shortcut to ``cy.visit( `/wp-admin/${page}` )``, not worth implementing at this time IMHO

**Comment by dinhtungdu** [comment]
Makes a lot of sense to me! Closing this PR.

**Pull request closed without merging**